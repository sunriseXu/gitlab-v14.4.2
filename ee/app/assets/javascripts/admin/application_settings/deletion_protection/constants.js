import { __, s__ } from '~/locale';

export const I18N_DELETION_PROTECTION = {
  heading: s__('DeletionSettings|Deletion protection'),
  helpText: s__(
    'DeletionSettings|Retention period that deleted groups and projects will remain restorable. Personal projects are always deleted immediately. Some groups can opt-out their projects.',
  ),
  learnMore: __('Learn more.'),
  keepDeleted: s__('DeletionSettings|Keep deleted'),
  deleteImmediately: s__('DeletionSettings|None, delete immediately'),
  for: __('for'),
  days: __('days'),
  groupsOnly: __('groups only'),
  groupsAndProjects: __('groups and projects'),
};
