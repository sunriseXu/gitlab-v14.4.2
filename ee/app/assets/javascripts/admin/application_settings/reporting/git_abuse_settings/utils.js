import { parseBoolean } from '~/lib/utils/common_utils';

export const parseFormProps = ({
  maxNumberOfRepositoryDownloads,
  maxNumberOfRepositoryDownloadsWithinTimePeriod,
  gitRateLimitUsersAllowlist,
  autoBanUserOnExcessiveProjectsDownload,
}) => ({
  maxNumberOfRepositoryDownloads: parseInt(maxNumberOfRepositoryDownloads, 10),
  maxNumberOfRepositoryDownloadsWithinTimePeriod: parseInt(
    maxNumberOfRepositoryDownloadsWithinTimePeriod,
    10,
  ),
  gitRateLimitUsersAllowlist: JSON.parse(gitRateLimitUsersAllowlist),
  autoBanUserOnExcessiveProjectsDownload: parseBoolean(autoBanUserOnExcessiveProjectsDownload),
});
