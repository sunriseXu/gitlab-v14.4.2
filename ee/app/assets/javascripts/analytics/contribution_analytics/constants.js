export const CHART_HEIGHT = 350;
export const INNER_CHART_HEIGHT = 200;
export const CHART_X_AXIS_ROTATE = 45;
export const CHART_X_AXIS_NAME_TOP_PADDING = 55;
