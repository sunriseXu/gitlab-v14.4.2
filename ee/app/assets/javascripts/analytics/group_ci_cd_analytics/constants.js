export const STAT_ERROR_PLACEHOLDER = '-';
export const formatWithUtc = true;
