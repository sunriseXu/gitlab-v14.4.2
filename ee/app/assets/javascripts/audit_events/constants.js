import { n__, s__, __ } from '~/locale';
import { OPERATOR_IS_ONLY } from '~/vue_shared/components/filtered_search_bar/constants';
import GroupToken from './components/tokens/group_token.vue';
import MemberToken from './components/tokens/member_token.vue';
import ProjectToken from './components/tokens/project_token.vue';
import UserToken from './components/tokens/user_token.vue';

const DEFAULT_TOKEN_OPTIONS = {
  operators: OPERATOR_IS_ONLY,
  unique: true,
};

// Due to the i18n eslint rule we can't have a capitalized string even if it is a case-aware URL param
/* eslint-disable @gitlab/require-i18n-strings */
export const ENTITY_TYPES = {
  USER: 'User',
  AUTHOR: 'Author',
  GROUP: 'Group',
  PROJECT: 'Project',
};
/* eslint-enable @gitlab/require-i18n-strings */

export const AUDIT_FILTER_CONFIGS = [
  {
    ...DEFAULT_TOKEN_OPTIONS,
    icon: 'user',
    title: s__('AuditLogs|User Events'),
    type: 'user',
    entityType: ENTITY_TYPES.USER,
    token: UserToken,
  },
  {
    ...DEFAULT_TOKEN_OPTIONS,
    icon: 'user',
    title: s__('AuditLogs|Member Events'),
    type: 'member',
    entityType: ENTITY_TYPES.AUTHOR,
    token: MemberToken,
  },
  {
    ...DEFAULT_TOKEN_OPTIONS,
    icon: 'bookmark',
    title: s__('AuditLogs|Project Events'),
    type: 'project',
    entityType: ENTITY_TYPES.PROJECT,
    token: ProjectToken,
  },
  {
    ...DEFAULT_TOKEN_OPTIONS,
    icon: 'group',
    title: s__('AuditLogs|Group Events'),
    type: 'group',
    entityType: ENTITY_TYPES.GROUP,
    token: GroupToken,
  },
];

export const AVAILABLE_TOKEN_TYPES = AUDIT_FILTER_CONFIGS.map((token) => token.type);

export const MAX_DATE_RANGE = 31;

export const SAME_DAY_OFFSET = 1;

// This creates a date with zero time, making it simpler to match to the query date values
export const CURRENT_DATE = new Date(new Date().toDateString());

export const AUDIT_EVENTS_TAB_TITLES = {
  LOG: s__('AuditLogs|Log'),
  STREAM: s__('AuditStreams|Streams'),
};

export const streamsLabel = (count) =>
  n__('AuditStreams|%d destination', 'AuditStreams|%d destinations', count);
export const ADD_STREAM = s__('AuditStreams|Add streaming destination');
export const ADD_STREAM_MESSAGE = s__('AuditStreams|Stream added successfully');
export const UPDATE_STREAM_MESSAGE = s__('AuditStreams|Stream updated successfully');
export const DELETE_STREAM_MESSAGE = s__('AuditStreams|Stream deleted successfully');

export const STREAM_ITEMS_I18N = {
  VERIFICATION_TOKEN_TOOLTIP: s__('AuditStreams|Verification token'),
  EDIT_BUTTON_LABEL: s__('AuditStreams|Edit %{link}'),
  DELETE_BUTTON_LABEL: s__('AuditStreams|Delete %{link}'),
  EDIT_BUTTON_TOOLTIP: __('Edit'),
  DELETE_BUTTON_TOOLTIP: __('Delete'),
};

export const ADD_STREAM_EDITOR_I18N = {
  WARNING_TITLE: s__('AuditStreams|Destinations receive all audit event data'),
  WARNING_CONTENT: s__(
    'AuditStreams|This could include sensitive information. Make sure you trust the destination endpoint.',
  ),
  DESTINATION_URL_LABEL: s__('AuditStreams|Destination URL'),
  DESTINATION_URL_PLACEHOLDER: 'https://api.gitlab.com',
  HEADERS_LABEL: s__('AuditStreams|Custom HTTP headers (optional)'),
  TABLE_COLUMN_NAME_LABEL: s__('AuditStreams|Header'),
  TABLE_COLUMN_VALUE_LABEL: s__('AuditStreams|Value'),
  TABLE_COLUMN_ACTIVE_LABEL: s__('AuditStreams|Active'),
  HEADER_INPUT_PLACEHOLDER: s__('AuditStreams|Add a custom header'),
  HEADER_INPUT_DUPLICATE_ERROR: s__('AuditStreams|A header with this name already exists.'),
  VALUE_INPUT_PLACEHOLDER: s__('AuditStreams|Add a custom value'),
  ADD_HEADER_ROW_BUTTON_TEXT: s__('AuditStreams|Add header'),
  ADD_HEADER_ROW_BUTTON_NAME: s__('AuditStreams|Add another custom header'),
  MAXIMUM_HEADERS_TEXT: s__('AuditStreams|Maximum of %{number} HTTP headers has been reached.'),
  REMOVE_BUTTON_LABEL: s__('AuditStreams|Remove custom header'),
  REMOVE_BUTTON_TOOLTIP: __('Remove'),
  ADD_BUTTON_TEXT: __('Add'),
  ADD_BUTTON_NAME: s__('AuditStreams|Add external stream destination'),
  SAVE_BUTTON_TEXT: __('Save'),
  SAVE_BUTTON_NAME: s__('AuditStreams|Save external stream destination'),
  CANCEL_BUTTON_TEXT: __('Cancel'),
  CANCEL_BUTTON_NAME: s__('AuditStreams|Cancel editing'),
};

export const AUDIT_STREAMS_EMPTY_STATE_I18N = {
  TITLE: s__('AuditStreams|Setup streaming for audit events'),
  DESCRIPTION_1: s__(
    'AuditStreams|Add an HTTP endpoint to manage audit logs in third-party systems.',
  ),
  DESCRIPTION_2: s__('AuditStreams|This is great for keeping everything one place.'),
};

export const AUDIT_STREAMS_NETWORK_ERRORS = {
  FETCHING_ERROR: s__(
    'AuditStreams|An error occurred when fetching external audit event streams. Please try it again',
  ),
  CREATING_ERROR: s__(
    'AuditStreams|An error occurred when creating external audit event stream destination. Please try it again.',
  ),
  UPDATING_ERROR: s__(
    'AuditStreams|An error occurred when updating external audit event stream destination. Please try it again.',
  ),
  DELETING_ERROR: s__(
    'AuditStreams|An error occurred when deleting external audit event stream destination. Please try it again.',
  ),
};

export const MAX_HEADERS = 20;
export const createBlankHeader = () => ({
  id: null,
  name: '',
  value: '',
  active: true,
  disabled: false,
  deletionDisabled: false,
  validationErrors: { name: '' },
});
