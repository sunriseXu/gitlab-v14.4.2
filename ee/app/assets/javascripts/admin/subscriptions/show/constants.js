import { __, s__ } from '~/locale';
import { PROMO_URL } from 'jh_else_ce/lib/utils/url_utility';

export const subscriptionMainTitle = s__('SuperSonics|Your subscription');
export const subscriptionActivationNotificationText = s__(
  `SuperSonics|Your subscription was successfully activated. You can see the details below.`,
);
export const subscriptionActivationFutureDatedNotificationTitle = s__(
  'SuperSonics|Your future dated license was successfully added',
);
export const subscriptionActivationFutureDatedNotificationMessage = s__(
  'SuperSonics|You have successfully added a license that activates on %{date}. Please see the subscription history table below for more details.',
);
export const subscriptionActivationInsertCode = __(
  "If you've purchased or renewed your subscription and have an activation code, please enter it below to start the activation process.",
);
export const howToActivateSubscription = s__(
  'SuperSonics|Learn how to %{linkStart}activate your subscription%{linkEnd}.',
);
export const subscriptionHistoryFailedTitle = s__('SuperSonics|Subscription unavailable');
export const subscriptionHistoryFailedMessage = s__(
  'SuperSonics|Your %{subscriptionEntryName} cannot be displayed at the moment. Please refresh the page to try again.',
);
export const currentSubscriptionsEntryName = s__('SuperSonics|current subscription');
export const pastSubscriptionsEntryName = s__('SuperSonics|past subscriptions');
export const futureSubscriptionsEntryName = s__('SuperSonics|future subscriptions');

export const cancelLabel = __('Cancel');
export const activateLabel = s__('AdminUsers|Activate');
export const activateSubscription = s__('SuperSonics|Activate subscription');
export const activateCloudLicense = s__('SuperSonics|Enter activation code');
export const noActiveSubscription = s__(`SuperSonics|You do not have an active subscription`);
export const subscriptionDetailsHeaderText = s__('SuperSonics|Subscription details');
export const licensedToHeaderText = s__('SuperSonics|Licensed to');
export const manageSubscriptionButtonText = s__('SuperSonics|Manage');
export const syncSubscriptionButtonText = s__('SuperSonics|Sync subscription details');
export const copySubscriptionIdButtonText = __('Copy');
export const licenseFileText = __('Legacy license');
export const onlineCloudLicenseText = s__('SuperSonics|Online license');
export const offlineCloudLicenseText = s__('SuperSonics|Offline license');
export const detailsLabels = {
  address: __('Address'),
  company: __('Company'),
  email: __('Email'),
  id: __('ID'),
  lastSync: __('Last Sync'),
  name: licensedToHeaderText,
  plan: __('Plan'),
  type: __('Type'),
  expiresAt: __('Renews'),
  startsAt: __('Started'),
};

export const subscriptionTable = {
  activatedAt: __('Activated on'),
  expiresOn: __('Expires on'),
  seats: __('Seats'),
  startsAt: __('Valid From'),
  title: __('Subscription History'),
  type: __('Type'),
};

export const subscriptionActivationForm = {
  activationCode: s__('SuperSonics|Activation code'),
  activationCodeFeedback: s__(
    'SuperSonics|The activation code should be a 24-character alphanumeric string',
  ),
  pasteActivationCode: s__('SuperSonics|Paste your activation code'),
  acceptTerms: s__(
    'SuperSonics|I agree that my use of the GitLab Software is subject to the Subscription Agreement located at the %{linkStart}Terms of Service%{linkEnd}, unless otherwise agreed to in writing with GitLab.',
  ),
  acceptTermsFeedback: s__('SuperSonics|Please agree to the Subscription Agreement'),
};

export const subscriptionSyncStatus = {
  SYNC_FAILURE: 'SYNC_FAILURE',
  SYNC_PENDING: 'SYNC_PENDING',
};

export const subscriptionTypes = {
  ONLINE_CLOUD: 'online_cloud',
  OFFLINE_CLOUD: 'offline_cloud',
  LEGACY_LICENSE: 'legacy_license',
};

export const trialCard = {
  title: s__('SuperSonics|Free trial'),
  description: s__(
    'SuperSonics|You can start a free trial of GitLab Ultimate without any obligation or payment details.',
  ),
  startTrial: s__('SuperSonics|Start free trial'),
};

export const buySubscriptionCard = {
  title: __('Subscription'),
  description: s__(
    'SuperSonics|Ready to get started? A GitLab plan is ideal for scaling organizations and for multi team usage.',
  ),
  buttonLabel: s__('SuperSonics|Buy subscription'),
};

export const SUBSCRIPTION_ACTIVATION_FAILURE_EVENT = 'subscription-activation-failure';
export const SUBSCRIPTION_ACTIVATION_SUCCESS_EVENT = 'subscription-activation-success';
export const SUBSCRIPTION_ACTIVATION_FINALIZED_EVENT = 'subscription-activation-finalized';

// Server-sent error messages (or regexes to match those messages) to detect error type
export const INVALID_ACTIVATION_CODE_SERVER_ERROR = 'invalid activation code';
export const EXPIRED_LICENSE_SERVER_ERROR = 'This license has already expired.';
export const SUBSCRIPTION_NOT_FOUND_SERVER_ERROR = 'without cloud compatible subscription';
export const SUBSCRIPTION_OVERAGES_SERVER_ERROR_REGEX = /(This GitLab installation currently has|During the year before this license started, this GitLab installation had) (\d+) active users?, exceeding this license's limit of (\d+) by (\d+) users?\. Please add a license for at least (\d+) users? or contact sales at https:\/\/about\.gitlab\.com\/sales\//;
export const SUBSCRIPTION_INSUFFICIENT_TRUE_UP_SERVER_ERROR_REGEX = /You have applied a True-up for (\d+) users? but you need one for (\d+) users?\. Please contact sales at https:\/\/about\.gitlab\.com\/sales\//;

export const CONNECTIVITY_ERROR = 'CONNECTIVITY_ERROR';
export const INVALID_CODE_ERROR = 'INVALID_CODE_ERROR';

export const supportLink = `${PROMO_URL}/support/#contact-support`;
export const subscriptionBannerTitle = s__('SuperSonics|Cloud licensing');
export const subscriptionBannerText = s__(
  "SuperSonics|Cloud licensing is now available. It's an easier way to activate instances and manage subscriptions. Read more about it in our %{blogPostLinkStart}blog post%{blogPostLinkEnd}. Activation codes are available in the %{portalLinkStart}Customers Portal%{portalLinkEnd}.",
);
export const subscriptionBannerBlogPostUrl =
  'https://about.gitlab.com/blog/2021/07/20/improved-billing-and-subscription-management/';
export const exportLicenseUsageBtnText = s__('SuperSonics|Export license usage file');

export const instanceHasFutureLicenseBanner = {
  title: s__('SuperSonics|You have a future dated license'),
  message: s__(
    'SuperSonics|You have added a license that activates on %{date}. Please see the subscription history table below for more details.',
  ),
};
