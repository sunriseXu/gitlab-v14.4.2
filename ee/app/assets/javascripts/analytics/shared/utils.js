import { dataVizBlue500, gray300 } from '@gitlab/ui/scss_to_js/scss_variables';
import { merge, cloneDeep } from 'lodash';
import { dateFormats } from '~/analytics/shared/constants';
import dateFormat from '~/lib/dateformat';
import { convertObjectPropsToCamelCase, parseBoolean } from '~/lib/utils/common_utils';
import { capitalizeFirstCharacter } from '~/lib/utils/text_utility';

export const formattedDate = (d) => dateFormat(d, dateFormats.defaultDate);

/**
 * Creates a value stream object from a dataset. Returns null if no valueStreamId is present.
 *
 * @param {Object} dataset - The raw value stream object
 * @returns {Object} - A value stream object
 */
export const buildValueStreamFromJson = (valueStream) => {
  const { id, name, is_custom: isCustom } = valueStream ? JSON.parse(valueStream) : {};
  return id ? { id, name, isCustom } : null;
};

/**
 * Creates an array of stage objects from a json string. Returns an empty array if no stages are present.
 *
 * @param {String} stages - JSON encoded array of stages
 * @returns {Array} - An array of stage objects
 */
const buildDefaultStagesFromJSON = (stages = '') => {
  if (!stages.length) return [];
  return JSON.parse(stages);
};

/**
 * Creates a group object from a dataset. Returns null if no groupId is present.
 *
 * @param {Object} dataset - The container's dataset
 * @returns {Object} - A group object
 */
export const buildGroupFromDataset = (dataset) => {
  const { groupId, groupName, groupFullPath, groupAvatarUrl, groupParentId } = dataset;

  if (groupId) {
    return {
      id: Number(groupId),
      name: groupName,
      full_path: groupFullPath,
      avatar_url: groupAvatarUrl,
      parent_id: groupParentId,
    };
  }

  return null;
};

/**
 * Creates a project object from a dataset. Returns null if no projectId is present.
 *
 * @param {Object} dataset - The container's dataset
 * @returns {Object} - A project object
 */
export const buildProjectFromDataset = (dataset) => {
  const { projectGid, projectName, projectPathWithNamespace, projectAvatarUrl } = dataset;

  if (projectGid) {
    return {
      id: projectGid,
      name: projectName,
      path_with_namespace: projectPathWithNamespace,
      avatar_url: projectAvatarUrl,
    };
  }

  return null;
};

/**
 * Creates a new date object without time zone conversion.
 *
 * We use this method instead of `new Date(date)`.
 * `new Date(date) will assume that the date string is UTC and it
 * ant return different date depending on the user's time zone.
 *
 * @param {String} date - Date string.
 * @returns {Date} - Date object.
 */
export const toLocalDate = (date) => {
  const dateParts = date.split('-');

  return new Date(dateParts[0], dateParts[1] - 1, dateParts[2]);
};

/**
 * Creates an array of project objects from a json string. Returns null if no projects are present.
 *
 * @param {String} data - JSON encoded array of projects
 * @returns {Array} - An array of project objects
 */
const buildProjectsFromJSON = (projects = '') => {
  if (!projects.length) return [];
  return JSON.parse(projects);
};

/**
 * Builds the initial data object for Value Stream Analytics with data loaded from the backend
 *
 * @param {Object} dataset - dataset object paseed to the frontend via data-* properties
 * @returns {Object} - The initial data to load the app with
 */
export const buildCycleAnalyticsInitialData = ({
  valueStream = null,
  groupId = null,
  createdBefore = null,
  createdAfter = null,
  projects = null,
  groupName = null,
  groupFullPath = null,
  groupParentId = null,
  groupAvatarUrl = null,
  labelsPath = '',
  milestonesPath = '',
  defaultStages = null,
  stage = null,
  aggregationEnabled = false,
  aggregationLastRunAt = null,
  aggregationNextRunAt = null,
} = {}) => ({
  selectedValueStream: buildValueStreamFromJson(valueStream),
  group: groupId
    ? convertObjectPropsToCamelCase(
        buildGroupFromDataset({
          groupId,
          groupName,
          groupFullPath,
          groupAvatarUrl,
          groupParentId,
        }),
      )
    : null,
  createdBefore: createdBefore ? toLocalDate(createdBefore) : null,
  createdAfter: createdAfter ? toLocalDate(createdAfter) : null,
  selectedProjects: projects
    ? buildProjectsFromJSON(projects).map(convertObjectPropsToCamelCase)
    : null,
  labelsPath,
  milestonesPath,
  defaultStageConfig: defaultStages
    ? buildDefaultStagesFromJSON(defaultStages).map(({ name, ...rest }) => ({
        ...convertObjectPropsToCamelCase(rest),
        name: capitalizeFirstCharacter(name),
      }))
    : [],
  stage: JSON.parse(stage),
  aggregation: {
    enabled: parseBoolean(aggregationEnabled),
    lastRunAt: aggregationLastRunAt,
    nextRunAt: aggregationNextRunAt,
  },
});

/**
 * Linearly interpolates between two values
 *
 * @param {Number} valueAtT0 The value at t = 0
 * @param {Number} valueAtT1 The value at t = 1
 * @param {Number} t The current value of t
 *
 * @returns {Number} The result of the linear interpolation.
 */
const lerp = (valueAtT0, valueAtT1, t) => {
  return valueAtT0 * (1 - t) + valueAtT1 * t;
};

/**
 * Builds a second series that visually represents the "no data" (i.e. "null")
 * data points, and returns a new series Array that includes both the "null"
 * and "non-null" data sets.
 * This function returns new series data and does not modify the original instance.
 *
 * @param {Array} seriesData The lead time series data that has already been processed
 * by the `apiDataToChartSeries` function above.
 * @returns {Array} A new series Array
 */
export const buildNullSeries = (seriesData, nullSeriesTitle) => {
  const nonNullSeries = cloneDeep(seriesData[0]);

  // Loop through the series data and build a list of all the "gaps". A "gap" is
  // a section of the data set that only include `null` values. Each gap object
  // includes the start and end indices and the start and end values of the gap.
  const seriesGaps = [];
  let currentGap = null;
  nonNullSeries.data.forEach(([, value], index) => {
    if (value == null && currentGap == null) {
      currentGap = {};

      if (index > 0) {
        currentGap.startIndex = index - 1;
        const [, previousValue] = nonNullSeries.data[index - 1];
        currentGap.startValue = previousValue;
      }

      seriesGaps.push(currentGap);
    } else if (value != null && currentGap != null) {
      currentGap.endIndex = index;
      currentGap.endValue = value;
      currentGap = null;
    }
  });

  // Create a copy of the non-null series, but with all the data point values set to `null`
  const nullSeriesData = nonNullSeries.data.map(([date]) => [date, null]);

  // Render each of the gaps to the "null" series. Values are determined by linearly
  // interpolating between the start and end values.
  seriesGaps.forEach((gap) => {
    const startIndex = gap.startIndex ?? 0;
    const startValue = gap.startValue ?? gap.endValue ?? 0;
    const endIndex = gap.endIndex ?? nonNullSeries.data.length - 1;
    const endValue = gap.endValue ?? gap.startValue ?? 0;

    for (let i = startIndex; i <= endIndex; i += 1) {
      const t = (i - startIndex) / (endIndex - startIndex);
      nullSeriesData[i][1] = lerp(startValue, endValue, t);
    }
  });

  merge(nonNullSeries, {
    showSymbol: true,
    showAllSymbol: true,
    symbolSize: 8,
    lineStyle: {
      color: dataVizBlue500,
    },
    areaStyle: {
      color: dataVizBlue500,
      opacity: 0,
    },
    itemStyle: {
      color: dataVizBlue500,
    },
  });

  const nullSeries = {
    name: nullSeriesTitle,
    data: nullSeriesData,
    lineStyle: {
      type: 'dashed',
      color: gray300,
    },
    areaStyle: {
      color: 'none',
    },
    itemStyle: {
      color: gray300,
    },
  };

  return [nullSeries, nonNullSeries];
};
