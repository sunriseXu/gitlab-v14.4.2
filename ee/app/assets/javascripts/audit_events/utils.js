import { parsePikadayDate, pikadayToString } from '~/lib/utils/datetime_utility';
import { getIdFromGraphQLId } from '~/graphql_shared/utils';
import {
  AVAILABLE_TOKEN_TYPES,
  AUDIT_FILTER_CONFIGS,
  ENTITY_TYPES,
  createBlankHeader,
} from './constants';
import { parseUsername, displayUsername } from './token_utils';

export const getTypeFromEntityType = (entityType) => {
  return AUDIT_FILTER_CONFIGS.find(
    ({ entityType: configEntityType }) => configEntityType === entityType,
  )?.type;
};

export const getEntityTypeFromType = (type) => {
  return AUDIT_FILTER_CONFIGS.find(({ type: configType }) => configType === type)?.entityType;
};

export const parseAuditEventSearchQuery = ({
  created_after: createdAfter,
  created_before: createdBefore,
  entity_type: entityType,
  entity_username: entityUsername,
  author_username: authorUsername,
  ...restOfParams
}) => ({
  ...restOfParams,
  created_after: createdAfter ? parsePikadayDate(createdAfter) : null,
  created_before: createdBefore ? parsePikadayDate(createdBefore) : null,
  entity_type: getTypeFromEntityType(entityType),
  entity_username: displayUsername(entityUsername),
  author_username: displayUsername(authorUsername),
});

export const createAuditEventSearchQuery = ({ filterValue, startDate, endDate, sortBy }) => {
  const entityValue = filterValue.find((value) => AVAILABLE_TOKEN_TYPES.includes(value.type));
  const entityType = getEntityTypeFromType(entityValue?.type);
  const filterData = entityValue?.value.data;

  const params = {
    created_after: startDate ? pikadayToString(startDate) : null,
    created_before: endDate ? pikadayToString(endDate) : null,
    sort: sortBy,
    entity_type: entityType,
    entity_id: null,
    entity_username: null,
    author_username: null,
    // When changing the search parameters, we should be resetting to the first page
    page: null,
  };

  switch (entityType) {
    case ENTITY_TYPES.USER:
      params.entity_username = parseUsername(filterData);
      break;
    case ENTITY_TYPES.AUTHOR:
      params.author_username = parseUsername(filterData);
      break;
    default:
      params.entity_id = filterData;
  }

  return params;
};

export const mapItemHeadersToFormData = (item, settings = {}) => {
  const headers = item?.headers?.nodes || [];

  return (
    headers
      .map(({ id, key, value }) => ({
        ...createBlankHeader(),
        id,
        name: key,
        value,
        ...settings,
      }))
      // Sort the headers so they appear in the order they were created
      // The GraphQL endpoint returns them in the reverse order of this
      .sort((a, b) => getIdFromGraphQLId(a.id) - getIdFromGraphQLId(b.id))
  );
};

export const mapAllMutationErrors = (mutations, name) => {
  return Promise.allSettled(mutations).then((results) => {
    const rejected = results.filter((r) => r.status === 'rejected').map((r) => r.reason);

    if (rejected.length > 0) {
      throw rejected[0];
    }

    return results
      .filter((r) => r.status === 'fulfilled')
      .map((r) => r.value.data[name].errors)
      .reduce((r, errors) => r.concat(errors), [])
      .filter(Boolean);
  });
};
