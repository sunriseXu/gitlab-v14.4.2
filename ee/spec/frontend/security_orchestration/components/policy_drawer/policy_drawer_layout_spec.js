import { GlSprintf, GlLink } from '@gitlab/ui';
import { getSourceUrl } from 'ee/security_orchestration/components/utils';
import { shallowMountExtended } from 'helpers/vue_test_utils_helper';
import PolicyDrawerLayout from 'ee/security_orchestration/components/policy_drawer/policy_drawer_layout.vue';
import {
  DEFAULT_DESCRIPTION_LABEL,
  ENABLED_LABEL,
  GROUP_TYPE_LABEL,
  INHERITED_LABEL,
  NOT_ENABLED_LABEL,
  PROJECT_TYPE_LABEL,
} from 'ee/security_orchestration/components/policy_drawer/constants';
import { NAMESPACE_TYPES } from 'ee/security_orchestration/constants';
import {
  mockGroupScanExecutionPolicy,
  mockProjectScanExecutionPolicy,
} from '../../mocks/mock_data';

describe('PolicyDrawerLayout component', () => {
  let wrapper;

  const DESCRIPTION = 'This policy enforces pipeline configuration to have a job with DAST scan';
  const TYPE = 'Scan Execution';

  const findCustomDescription = () => wrapper.findByTestId('custom-description-text');
  const findDefaultDescription = () => wrapper.findByTestId('default-description-text');
  const findEnabledText = () => wrapper.findByTestId('enabled-status-text');
  const findNotEnabledText = () => wrapper.findByTestId('not-enabled-status-text');
  const findSourceSection = () => wrapper.findByTestId('policy-source');
  const findSprintf = () => wrapper.findComponent(GlSprintf);
  const findLink = () => wrapper.findComponent(GlLink);

  const componentStatusText = (status) => (status ? 'does' : 'does not');

  const factory = ({ propsData = {}, provide = { namespaceType: NAMESPACE_TYPES.PROJECT } }) => {
    wrapper = shallowMountExtended(PolicyDrawerLayout, {
      provide,
      propsData: {
        type: TYPE,
        ...propsData,
      },
      scopedSlots: {
        summary: `<span data-testid="summary-text">Summary</span>`,
      },
      stubs: {
        GlSprintf,
      },
    });
  };

  afterEach(() => {
    wrapper.destroy();
  });

  describe.each`
    context                 | propsData                                                               | enabled  | hasDescription
    ${'enabled policy'}     | ${{ policy: mockProjectScanExecutionPolicy, description: DESCRIPTION }} | ${true}  | ${true}
    ${'not enabled policy'} | ${{ policy: { ...mockProjectScanExecutionPolicy, enabled: false } }}    | ${false} | ${false}
  `('$context', ({ enabled, hasDescription, propsData }) => {
    beforeEach(() => {
      factory({ propsData });
    });

    it.each`
      component                | status                                  | finder                    | exists             | text
      ${'custom description'}  | ${componentStatusText(hasDescription)}  | ${findCustomDescription}  | ${hasDescription}  | ${DESCRIPTION}
      ${'default description'} | ${componentStatusText(!hasDescription)} | ${findDefaultDescription} | ${!hasDescription} | ${DEFAULT_DESCRIPTION_LABEL}
      ${'enabled text'}        | ${componentStatusText(enabled)}         | ${findEnabledText}        | ${enabled}         | ${ENABLED_LABEL}
      ${'not enabled text'}    | ${componentStatusText(!enabled)}        | ${findNotEnabledText}     | ${!enabled}        | ${NOT_ENABLED_LABEL}
    `('$status render the $component', ({ exists, finder, text }) => {
      const component = finder();
      expect(component.exists()).toBe(exists);
      if (exists) {
        expect(component.text()).toBe(text);
      }
    });

    it('matches the snapshots', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('source field', () => {
    it('displays correctly for a project-level policy being displayed on a project', () => {
      factory({ propsData: { policy: mockProjectScanExecutionPolicy } });
      expect(findSourceSection().text()).toBe(PROJECT_TYPE_LABEL);
    });

    it('displays correctly for a group-level policy being displayed on a project', () => {
      factory({ propsData: { policy: mockGroupScanExecutionPolicy } });
      expect(findSprintf().text()).toMatchInterpolatedText(INHERITED_LABEL);
      expect(findLink().attributes('href')).toBe(
        getSourceUrl(mockGroupScanExecutionPolicy.source.namespace.fullPath),
      );
    });

    it('displays correctly for a group-level policy being displayed on a group', () => {
      factory({
        propsData: { policy: mockProjectScanExecutionPolicy },
        provide: { namespaceType: NAMESPACE_TYPES.GROUP },
      });
      expect(findSourceSection().text()).toBe(GROUP_TYPE_LABEL);
    });
  });
});
