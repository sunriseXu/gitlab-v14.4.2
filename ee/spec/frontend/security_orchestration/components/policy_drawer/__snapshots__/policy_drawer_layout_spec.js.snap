// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PolicyDrawerLayout component enabled policy matches the snapshots 1`] = `
<div>
  <div
    data-testid="policy-summary"
  >
    <span
      data-testid="summary-text"
    >
      Summary
    </span>
  </div>
   
  <policy-info-row-stub
    data-testid="policy-type"
    label="Policy Type"
  >
    
    Scan Execution
  
  </policy-info-row-stub>
   
  <policy-info-row-stub
    label="Description"
  >
    <div
      data-testid="custom-description-text"
    >
      
      This policy enforces pipeline configuration to have a job with DAST scan
    
    </div>
  </policy-info-row-stub>
   
  <policy-info-row-stub
    label="Source"
  >
    <div
      data-testid="policy-source"
    >
      <span>
        This is a project-level policy
      </span>
    </div>
  </policy-info-row-stub>
   
  <policy-info-row-stub
    label="Status"
  >
    <div
      class="gl-text-green-500"
      data-testid="enabled-status-text"
    >
      <gl-icon-stub
        class="gl-mr-3"
        name="check-circle-filled"
        size="16"
      />
      Enabled
    
    </div>
  </policy-info-row-stub>
</div>
`;

exports[`PolicyDrawerLayout component not enabled policy matches the snapshots 1`] = `
<div>
  <div
    data-testid="policy-summary"
  >
    <span
      data-testid="summary-text"
    >
      Summary
    </span>
  </div>
   
  <policy-info-row-stub
    data-testid="policy-type"
    label="Policy Type"
  >
    
    Scan Execution
  
  </policy-info-row-stub>
   
  <policy-info-row-stub
    label="Description"
  >
    <div
      class="gl-text-gray-500"
      data-testid="default-description-text"
    >
      
      No description
    
    </div>
  </policy-info-row-stub>
   
  <policy-info-row-stub
    label="Source"
  >
    <div
      data-testid="policy-source"
    >
      <span>
        This is a project-level policy
      </span>
    </div>
  </policy-info-row-stub>
   
  <policy-info-row-stub
    label="Status"
  >
    <div
      class="gl-text-gray-500"
      data-testid="not-enabled-status-text"
    >
      
      Not enabled
    
    </div>
  </policy-info-row-stub>
</div>
`;
